### PR TITLE
Add Nemotron ASR Streaming

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

Should be merged before: #46331

original checkpoints:
- [nvidia/nemotron-speech-streaming-en-0.6b](https://huggingface.co/nvidia/nemotron-speech-streaming-en-0.6b)

### CODE REVIEW

File: src/transformers/models/nemotron_asr_streaming/generation_nemotron_asr_streaming.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+
+class NemotronAsrStreamingRNNTDecoderCache(ParakeetRNNTDecoderCache): ...
+
+
+class NemotronAsrStreamingGenerateOutput(ParakeetRNNTGenerateOutput): ...
+
+
+class NemotronAsrStreamingGenerationMixin(ParakeetRNNTGenerationMixin):
+    """Generation mixin for NemotronAsrStreaming RNN-T models.
```
Comment (eustlb) [inline src/transformers/models/nemotron_asr_streaming/generation_nemotron_asr_streaming.py:34]: Not 100% perfect in terms of readibility/ simplicity BUT that's okay: I did not want to invest to much time there I'll refactor soon, by adding factorizing methods to generate (see [this](https://github.com/huggingface/transformers/blob/13a30429735d0f0b91e01b3d5b96d84c88197419/src/transformers/models/parakeet/generation_parakeet.py#L243) former comment)

**Comment by eustlb** [comment]
@run-slow: nemotron_asr_streaming

File: docs/source/en/model_doc/nemotron_asr_streaming.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+model_id = "nvidia/nemotron-speech-streaming-en-0.6b"
+revision = "refs/pr/17"
+processor = AutoProcessor.from_pretrained(model_id, revision=revision)
+model = AutoModelForRNNT.from_pretrained(model_id, revision=revision, device_map="auto")
+
+audio = load_audio(
+    "https://huggingface.co/datasets/hf-internal-testing/dummy-audio-samples/resolve/main/bcn_weather.mp3",
+    sampling_rate=processor.feature_extractor.sampling_rate,
+)
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/nemotron_asr_streaming.md:59]: doesn't processor support url?

File: docs/source/en/model_doc/nemotron_asr_streaming.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    "input_features": input_features_generator(),
+    "streamer": streamer,
+}
+thread = Thread(target=model.generate, kwargs=generate_kwargs)
+thread.start()
+
+# Iterate over the streamer to get text chunks as they are generated
+print("Model output (streaming):", end=" ", flush=True)
+for text_chunk in streamer:
+    print(text_chunk, end="", flush=True)
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/nemotron_asr_streaming.md:139]: super neeeeat!

Comment (ArthurZucker) [reply docs/source/en/model_doc/nemotron_asr_streaming.md:139]: Is there a way to have a "simpler" API with something like 
```python
from transformers import AutoModelForCausalLM, AutoTokenizer, AudioIteratorStreamer
from threading import Thread

processor = AutoProcessor.from_pretrained(model_id, revision=revision)
model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2")
streamer = AudioIteratorStreamer(processor, num_lookahead_tokens=6)

generation_kwargs = dict(**audio, streamer=streamer, max_new_tokens=50)
thread = Thread(target=model.generate, kwargs=generation_kwargs)
thread.start()

for text_chunk in streamer:
    print(text_chunk, end="", flush=True)
```

That would match the way we do it for text? WDYT?

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 4]
```diff
@@ -0,0 +1,1329 @@
...
+        return self.kernel_size[0] - 1
+
+    @cached_property
+    def time_pad(self):
+        return (self.kernel_size[0] - 1, self.stride[0] - 1)
+
+    @cached_property
+    def freq_pad(self):
+        return (self.kernel_size[1] - 1, self.stride[1] - 1)
+
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: is it that important to cache this?

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 5]
```diff
@@ -0,0 +1,1329 @@
...
+        )
+
+        relative_key_states = self.relative_k_proj(position_embeddings)
+        relative_key_states = relative_key_states.view(batch_size, -1, self.config.num_attention_heads, self.head_dim)
+
+        # terms (b) and (d) — slice to total_key_length to cover cache + current chunk
+        matrix_bd = query_states_with_bias_v @ relative_key_states.permute(0, 2, 3, 1)
+        matrix_bd = self._rel_shift(matrix_bd)
+        matrix_bd = matrix_bd[..., :total_key_length]
+        matrix_bd = matrix_bd * self.scaling
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:849]: what does this represent? is this like ALIBI  bias?

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 6]
```diff
@@ -0,0 +1,1329 @@
...
+
+        if not is_torchdynamo_compiling():
+            torch._dynamo.mark_static_address(self.cache)
+
+        self.is_first_chunk = True
+        self.is_initialized = True
+
+    def update(self, hidden_states, conv_module=None):
+        if not self.is_initialized and conv_module is not None:
+            self.lazy_initialization(hidden_states, conv_module)
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:594]: i remember something about just using config and not module in a previous pr

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 7]
```diff
@@ -0,0 +1,1329 @@
...
+            self.layers.append(nn.Conv2d(self.channels, self.channels, kernel_size=1))
+            # activation
+            self.layers.append(nn.ReLU())
+
+        # Compute output freq length by simulating the conv chain with the actual padding applied.
+        pad_left, pad_right = self.kernel_size - 1, self.stride - 1
+        out_length = config.num_mel_bins
+        total_pad = pad_left + pad_right
+        for _ in range(self.num_layers):
+            out_length = (out_length + total_pad - self.kernel_size) // self.stride + 1
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: mmm no we usually do these computation in the config

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 8]
```diff
@@ -0,0 +1,1329 @@
...
+        streaming = padding_cache is not None
+
+        for layer in self.layers:
+            if isinstance(layer, NemotronAsrStreamingEncoderCausalConv2D):
+                hidden_states = layer(hidden_states, padding_cache=padding_cache)
+            else:
+                hidden_states = layer(hidden_states)
+
+            # mask the hidden states
+            if isinstance(layer, nn.Conv2d) and attention_mask is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: not a fan of the many codepath. EX: the cache is only for non `NemotronAsrStreamingEncoderCausalConv2D`, 
```python
            self.layers.append(
                NemotronAsrStreamingEncoderCausalConv2D(
                    self.channels,
                    self.channels,
                    kernel_size=self.kernel_size,
                    stride=self.stride,
                    groups=self.channels,
                    cache_key=f"subsampling.{idx + 1}",
                )
            )
            # pointwise conv
            self.layers.append(nn.Conv2d(self.channels, self.channels, kernel_size=1))
            # activation
            self.layers.append(nn.ReLU())
```

is something we usally avoid

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 9]
```diff
@@ -0,0 +1,1329 @@
...
+        hidden_states = self.pointwise_conv1(hidden_states)
+        hidden_states = nn.functional.glu(hidden_states, dim=1)
+
+        # Apply padding mask before convolution. The mask may be 3D (B, T_q, T_k) for offline or
+        # streaming use; reduce over the key dim to get a per-query padding indicator.
+        if attention_mask is not None:
+            if attention_mask.dtype == torch.bool:
+                all_masked_rows = torch.all(~attention_mask, dim=-1)
+            else:
+                all_masked_rows = torch.all(attention_mask == 0.0, dim=-1)
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: seems like the logical not is done often on the mask no? (each layer)

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 10]
```diff
@@ -0,0 +1,1329 @@
...
+        input_shape = hidden_states.shape[:-1]
+        batch_size, seq_length = input_shape
+        hidden_shape = (batch_size, seq_length, -1, self.head_dim)
+
+        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+
+        if past_key_values is not None:
+            key_states, value_states = past_key_values.update(key_states, value_states, self.layer_idx)
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:827]:
```suggestion
        if past_key_values is not None: # DIFF WITH PARAKEET?
            key_states, value_states = past_key_values.update(key_states, value_states, self.layer_idx)
```

Comment (ArthurZucker) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:827]: cache used because of the streaming nature I suppose?

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 11]
```diff
@@ -0,0 +1,1329 @@
...
+
+        hidden_states = self.activation(hidden_states)
+        hidden_states = self.pointwise_conv2(hidden_states)
+        hidden_states = hidden_states.transpose(1, 2)  # (B, T, C)
+
+        return hidden_states
+
+
+class NemotronAsrStreamingEncoderAttention(ParakeetEncoderAttention):
+    """Multi-head attention with relative positional encoding. See section 3.3 of https://huggingface.co/papers/1901.02860."""
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: please update to help isolating diff

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 12]
```diff
@@ -0,0 +1,1329 @@
...
+            out_length = (out_length + total_pad - self.kernel_size) // self.stride + 1
+        self.linear = nn.Linear(config.subsampling_conv_channels * out_length, config.hidden_size, bias=True)
+
+    def _get_output_length(self, input_lengths: torch.Tensor, conv_layer: nn.Conv2d, streaming: bool = False):
+        if not (hasattr(conv_layer, "stride") and conv_layer.stride != (1, 1)):
+            return input_lengths
+
+        kernel_size = conv_layer.kernel_size[0]
+        stride = conv_layer.stride[0]
+        if isinstance(conv_layer, NemotronAsrStreamingEncoderCausalConv2D):
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: we should get rid of these if else everywhere (just putting NemotronAsrStreamingEncoderCausalConv2D alone and + Conv2d + RElu alon? IDK its a bit weird to not group in 3

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py [hunk 13]
```diff
@@ -0,0 +1,1329 @@
...
+                **kwargs,
+            )
+        elif not isinstance(encoder_outputs, NemotronAsrStreamingEncoderModelOutput):
+            encoder_outputs = NemotronAsrStreamingEncoderModelOutput(
+                last_hidden_state=encoder_outputs[0] if len(encoder_outputs) > 0 else None,
+                pooler_output=encoder_outputs[1] if len(encoder_outputs) > 1 else None,
+                hidden_states=encoder_outputs[2] if len(encoder_outputs) > 2 else None,
+                attentions=encoder_outputs[3] if len(encoder_outputs) > 3 else None,
+                attention_mask=encoder_outputs[4] if len(encoder_outputs) > 4 else None,
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: nope shuld alays be dict here never anything else with correct output recorders

**Review by ArthurZucker — APPROVED** [review]
Ty! I think we can improve upon parakeet / gemma4 ! Let s also add a streamer no?

File: docs/source/en/model_doc/nemotron_asr_streaming.md (hunk 3 above)
Comment (eustlb) [reply docs/source/en/model_doc/nemotron_asr_streaming.md:139]: that EXACTLY the API I've drafted ahah !! coming in a subsequent PR though

File: docs/source/en/model_doc/nemotron_asr_streaming.md (hunk 2 above)
Comment (eustlb) [reply docs/source/en/model_doc/nemotron_asr_streaming.md:59]: not via __call__, only via chat_templates

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 5 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:849]: nope it's different, this is transformers XL attention, already implemented in parakeet, but just with the addition of past_key_values

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 10 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:827]: indeed!

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 6 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py:594]: actually in a previous PR (voxtral_realtime) we moved to the config only to module approach. Anyway now that multiple models use this approach it's the good moment to standardize, so we can discuss this in a subsequent PR that we'll move causalconv cache from modulars

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 4 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: yep you're right this is over-engineered, removed the `cached_properties`

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 7 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: agree! added in [ca3c035](https://github.com/huggingface/transformers/pull/46332/commits/ca3c035fa7198a91bde5eadf2739480f240f72fc)

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 8 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: agree! added in [ca3c035](https://github.com/huggingface/transformers/pull/46332/commits/ca3c035fa7198a91bde5eadf2739480f240f72fc)

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 11 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: done 👌

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 12 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: agree! added in [ca3c035](https://github.com/huggingface/transformers/pull/46332/commits/ca3c035fa7198a91bde5eadf2739480f240f72fc)

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 13 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: absolutely, thanks for catching

**Comment by eustlb** [comment]
run-slow: nemotron_asr_streaming

File: src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py (hunk 9 above)
Comment (eustlb) [reply src/transformers/models/nemotron_asr_streaming/modular_nemotron_asr_streaming.py]: yep! added all_masked_rows for that

**Comment by eustlb** [comment]
slow test passing locally (run-slow not working for some reason)! Ready to merge 🎉
